File: backend/snapstudio_core/compatibility.py
"""Compatibility Doctor — read-only diagnostics for common Snapmaker U1 /
Snorca / Orca project problems.

STRICTLY READ-ONLY. This module opens the 3MF, inspects
`Metadata/project_settings.config`, and returns findings. It NEVER writes,
mutates a profile, edits g-code, or touches the 3MF writer. It does not claim to
fix anything — it explains the problem and the safest next step.

Detects:
  A. Invalid 3MF/profile values (out-of-range slicer settings).
  B. Non-U1 / suspicious profile signature.
  C. Relative extrusion without a G92 E0 layer reset.
"""
from __future__ import annotations

from .container import ThreeMF
from .config_io import load_project_settings

SETTINGS = "Metadata/project_settings.config"

# key -> (min, max) inclusive valid range. Values below min / above max are flagged.
_RANGE_RULES: list[tuple[str, int, int]] = [
    ("prime_tower_brim_width", 0, 2147483647),
    ("raft_first_layer_expansion", 0, 2147483647),
    ("tree_support_wall_count", 0, 2),
    ("solid_infill_filament", 1, 2147483647),
    ("sparse_infill_filament", 1, 2147483647),
    ("wall_filament", 1, 2147483647),
]


def _first(v):
    """Orca stores most values as strings or single-element lists; normalize."""
    if isinstance(v, list):
        return v[0] if v else None
    return v


def _num(cfg: dict, key: str):
    """Return a float for a numeric setting, or None if absent/non-numeric."""
    if key not in cfg:
        return None
    raw = _first(cfg[key])
    if raw is None:
        return None
    try:
        return float(str(raw).strip())
    except (ValueError, TypeError):
        return None


def _truthy(cfg: dict, key: str) -> bool:
    raw = _first(cfg.get(key))
    if raw is None:
        return False
    return str(raw).strip().lower() in {"1", "true", "yes", "on"}


def _text(cfg: dict, key: str) -> str:
    raw = _first(cfg.get(key))
    return "" if raw is None else str(raw)


def _finding(fid, severity, title, explanation, setting_path, action, evidence) -> dict:
    return {
        "id": fid,
        "severity": severity,           # "error" | "warning" | "info"
        "title": title,
        "explanation": explanation,
        "setting_path": setting_path,
        "suggested_action": action,
        "evidence": evidence,
    }


def check(path: str) -> dict:
    """Read-only compatibility check. Returns {findings, summary, recommendation}."""
    tm = ThreeMF.open(path)
    if not tm.has_part(SETTINGS):
        return {
            "findings": [],
            "summary": "No project settings found to check — this looks like a bare "
                       "model (e.g. an STL) rather than a slicer project.",
            "recommendation": "Open it in a clean Snapmaker U1 project to slice.",
        }

    cfg = load_project_settings(tm.read_part(SETTINGS))
    findings: list[dict] = []

    # A. invalid / out-of-range values
    for key, lo, hi in _RANGE_RULES:
        val = _num(cfg, key)
        if val is None:
            continue
        if val < lo or val > hi:
            findings.append(_finding(
                f"value.{key}", "error",
                f"Invalid value for {key}",
                "This file likely carries settings from another printer or a stale "
                "profile (often a MakerWorld/Bambu-style project). The model may be "
                "fine; the project settings are the problem.",
                f"{SETTINGS} -> {key}",
                "Import the model into a clean Snapmaker U1 project instead of opening "
                "the foreign 3MF as the active project.",
                f"{key} = {_first(cfg[key])} (valid range [{lo},{hi}])",
            ))

    # B. non-U1 / suspicious profile signature
    model = _text(cfg, "printer_model")
    settings_id = _text(cfg, "printer_settings_id")
    sig = f"{model} {settings_id}".lower()
    if model or settings_id:
        if "u1" not in sig or "snapmaker" not in sig:
            findings.append(_finding(
                "profile.not_u1", "warning",
                "Profile may not be a Snapmaker U1 profile",
                "This file likely carries settings from another printer. Slicing it "
                "as-is can produce wrong results.",
                f"{SETTINGS} -> printer_model / printer_settings_id",
                "Make sure the active printer profile is Snapmaker U1, or import the "
                "model into a clean Snapmaker U1 project.",
                f"printer_model={model or '(none)'}; printer_settings_id={settings_id or '(none)'}",
            ))

    # C. relative extrusion without a G92 E0 layer reset
    if _truthy(cfg, "use_relative_e_distances"):
        layer_gcode = (_text(cfg, "layer_change_gcode") + "\n"
                       + _text(cfg, "before_layer_change_gcode")).upper()
        if "G92 E0" not in layer_gcode:
            findings.append(_finding(
                "extrusion.relative_no_reset", "warning",
                "Relative extrusion without a per-layer reset",
                "Your profile uses relative extrusion but may not reset the extruder "
                "position each layer. Newer Orca/Snorca versions are stricter about this.",
                f"{SETTINGS} -> layer_change_gcode / before_layer_change_gcode",
                "Switch back to the official Snapmaker U1 profile (or reset the profile). "
                "Only add G92 E0 to the layer-change g-code if you understand the setting.",
                "use_relative_e_distances is enabled; no 'G92 E0' found in layer-change g-code",
            ))

    errors = sum(1 for f in findings if f["severity"] == "error")
    warns = sum(1 for f in findings if f["severity"] == "warning")
    if not findings:
        summary = "No known U1 compatibility issues detected in this project's settings."
        recommendation = "Looks compatible with Snapmaker U1 — slice as usual."
    else:
        summary = f"Found {errors} invalid-value issue(s) and {warns} warning(s)."
        recommendation = (
            "Prefer importing the model into a clean Snapmaker U1 project rather than "
            "opening this foreign 3MF as the active project. Studio diagnoses these "
            "issues read-only; it does not change your file."
        )

    return {"findings": findings, "summary": summary, "recommendation": recommendation}

File: desktop/src/lib/compatibility.ts
// Pure presentation logic + copy for the Compatibility Doctor. No JSX so the
// beginner wording and finding-ordering are unit-testable. Read-only feature:
// Studio diagnoses and explains; it never claims to auto-fix.
import type { CompatibilityFinding } from "@/api";

export const COMPAT_COPY = {
  title: "Compatibility Doctor",
  subtitle: "Checks a 3MF for common Snapmaker U1 and Orca slicer problems — read-only.",
  cleanTitle: "No known compatibility issues",
  intro: "Open a 3MF project and Studio reads its settings to spot issues before you slice.",
  // Explicitly NOT an auto-fix claim:
  readOnlyNote: "Studio diagnoses these issues read-only. It does not change your file or fix them for you.",
  beginnerForeign: "This file likely carries settings from another printer.",
  beginnerImport: "Import the model into a clean Snapmaker U1 project instead of opening the foreign 3MF as the active project.",
  beginnerRelE: "Your profile uses relative extrusion but may not reset the extruder each layer.",
} as const;

const RANK: Record<string, number> = { error: 0, warning: 1, info: 2 };

/** Errors first, then warnings, then info; stable within a severity. */
export function sortFindings(findings: CompatibilityFinding[]): CompatibilityFinding[] {
  return [...findings].sort((a, b) => (RANK[a.severity] ?? 9) - (RANK[b.severity] ?? 9));
}

export function severityLabel(sev: string): string {
  if (sev === "error") return "Needs attention";
  if (sev === "warning") return "Heads up";
  return "Note";
}

/** Brand status token for a severity (used for swatch/border colour). */
export function severityToken(sev: string): string {
  if (sev === "error") return "--stage-input";       // risk-ish
  if (sev === "warning") return "--doctor-cost";      // amber-ish
  return "--doctor-project";
}

export interface CompatCounts { errors: number; warnings: number; info: number; total: number; }
export function countFindings(findings: CompatibilityFinding[]): CompatCounts {
  const c: CompatCounts = { errors: 0, warnings: 0, info: 0, total: findings.length };
  for (const f of findings) {
    if (f.severity === "error") c.errors++;
    else if (f.severity === "warning") c.warnings++;
    else c.info++;
  }
  return c;
}
